Add average uptime calculation

thaarok opened the pull request:
* Updated and refactored version of #83
* Use Q1.30 number representation for the normalised uptime (fits into uint32)
* Used math [lemma for the average](https://docs.google.com/document/d/1kENdl4XBtBV7Zt7VmJ9Nvaq-gsinPrCbfog_CxtFHgI/)
* Used unsigned integers only to simplify checks/verification
* Disabled too inactive validators included but disabled by default. (I would prefer to include it into the audit.)

b-scholz commented on contracts/sfc/ConstantsManager.sol:
```diff
@@ -153,4 +161,22 @@ contract ConstantsManager is Ownable {
+
+    function updateMinAverageUptime(uint32 v) external virtual onlyOwner {
+        if (v > 966367641) {
+            // 0.9 in Q1.30
```
Fix to UQ1.31.

b-scholz commented on contracts/sfc/SFC.sol:
```diff
@@ -68,13 +69,25 @@ contract SFC is Initializable, Ownable, Version {
 
+    // data structure to compute average uptime for each active validator
+    struct AverageUptime {
+        // average uptime ratio as a value between 0 and (1 << 30)
```
1<<31 (not 1<<30).

b-scholz commented on contracts/sfc/SFC.sol:
```diff
@@ -901,6 +920,70 @@ contract SFC is Initializable, Ownable, Version {
+    ) internal {
+        for (uint256 i = 0; i < validatorIDs.length; i++) {
+            uint256 validatorID = validatorIDs[i];
+            // compute normalised uptime as a percentage in the Q1.30 fixed-point format
```
Q1.30 => UQ1.31

b-scholz commented on contracts/sfc/SFC.sol:
```diff
@@ -901,6 +920,70 @@ contract SFC is Initializable, Ownable, Version {
+        for (uint256 i = 0; i < validatorIDs.length; i++) {
+            uint256 validatorID = validatorIDs[i];
+            // compute normalised uptime as a percentage in the Q1.30 fixed-point format
+            uint256 normalisedUptime = (uptimes[i] * (1 << 30)) / epochDuration;
```
1 << 30 =>  1 << 31 (Perhaps use a constant somewhere for the fixpoint such as `const UQ1.0 = 1 << 31;` and use the symbol instead of 1<<31.

b-scholz commented on contracts/sfc/SFC.sol:
```diff
@@ -901,6 +920,70 @@ contract SFC is Initializable, Ownable, Version {
+            // compute normalised uptime as a percentage in the Q1.30 fixed-point format
+            uint256 normalisedUptime = (uptimes[i] * (1 << 30)) / epochDuration;
+            if (normalisedUptime > 1 << 30) {
+                normalisedUptime = 1 << 30;
```
1 << 30 => 1 << 31

b-scholz commented on contracts/sfc/SFC.sol:
```diff
@@ -901,6 +920,70 @@ contract SFC is Initializable, Ownable, Version {
+        uint64 tmp = (n - 1) * uint64(prev.averageUptime) + uint64(newValue);
+
+        // apply remainder from the division in the previous iteration to reduce error
+        if (prev.remainder != 0) {
```
Remove the condition and keep the statement inside the if-statement only.

thaarok replied:
with using Decimal.unit(), the precision should be much higher, so the remainder should be meaningless anymore - removed.

b-scholz commented on contracts/sfc/SFC.sol:
```diff
@@ -901,6 +920,70 @@ contract SFC is Initializable, Ownable, Version {
+
+        // apply remainder from the division in the previous iteration to reduce error
+        if (prev.remainder != 0) {
+            tmp += (n * prev.remainder) / (n - 1);
```
Rewrite to `tmp += prev.remainder.` The reason is that if the threshold for the epoch window is reached, it should be prev.remainder only. For smaller n, the term n / (n-1) will converge quickly to zero (i.e., 1 / (1-1/n) ) and hence the scaling will have only little effect.

thaarok replied:
with using Decimal.unit(), the precision should be much higher, so the remainder should be meaningless anymore - removed.

thaarok replied:
fixed

b-scholz commented on contracts/sfc/SFC.sol:
```diff
@@ -901,6 +920,70 @@ contract SFC is Initializable, Ownable, Version {
+        cur.averageUptime = uint32(tmp / n);
+        cur.remainder = uint32(tmp % n);
+
+        if (cur.averageUptime > 1 << 30) {
```
1<<30 => 1 << 31

b-scholz commented on contracts/sfc/SFC.sol:
```diff
@@ -901,6 +920,70 @@ contract SFC is Initializable, Ownable, Version {
+        cur.remainder = uint32(tmp % n);
+
+        if (cur.averageUptime > 1 << 30) {
+            cur.averageUptime = 1 << 30;
```
1<<30 => 1 << 31

thaarok commented on contracts/sfc/ConstantsManager.sol:
```diff
@@ -153,4 +161,22 @@ contract ConstantsManager is Ownable {
+
+    function updateAverageUptimeEpochsThreshold(uint32 v) external virtual onlyOwner {
+        if (v < 10) {
+            revert ValueTooSmall();
```
I don't think we should fix and mention the epoch duration in SFC source, as it depends on the network configuration. We may use very different epochs configuration in the future.

b-scholz replied:
ok.

thaarok commented on contracts/sfc/ConstantsManager.sol:
```diff
@@ -29,6 +29,14 @@ contract ConstantsManager is Ownable {
+    // Is also the minimum number of epochs necessary for deactivation of offline validators.
+    uint32 public averageUptimeEpochsThreshold;
+
+    // Minimum average uptime in Q1.30 format; acceptable bounds [0,0.9]
```
Actually we have already Decimal.unit() in the SFC for this purpose.
Reworked to use Decimal.unit() instead.

thaarok commented:
Rewritten to use Unit.decimal() (1e18) instead of (1<<30) to keep fixed-point representation consistent across usages in the SFC.
As this provides much higher precision of the calculation, I have also removed the remainder error correction to keep the code simple.
